### DISCUSSION TITLE
Google Colab Inference Error

### DISCUSSION THREAD

**Discussion by jramborger78** [talmolab/sleap#1401] [Help!]
Hi there!

First wanted to say this software is amazing, kudos to you all.

Secondly, I keep running into an error on Google Colab when I try to run inference. I have been using just my basic laptop with the GUI to set up labels and export a training package, I then follow the Google Colab to a T (changing the instance to "single" as I am with one rat) and every time it returns this error whether I start a fresh notebook and put in the interface commands manually or edit my features to the names in the Colab demo:

INFO:numexpr.utils:NumExpr defaulting to 2 threads.
Started inference at: 2023-07-21 16:08:09.412217
Args:
{
│   'data_path': 'colab.pkg.slp',
│   'models': ['models/230721_083104.single_instance'],
│   'frames': '',
│   'only_labeled_frames': False,
│   'only_suggested_frames': True,
│   'output': 'colab.predicted_suggestions.slp',
│   'no_empty_frames': False,
│   'verbosity': 'rich',
│   'video.dataset': None,
│   'video.input_format': 'channels_last',
│   'video.index': '',
│   'cpu': False,
│   'first_gpu': False,
│   'last_gpu': False,
│   'gpu': 'auto',
│   'max_edge_length_ratio': 0.25,
│   'dist_penalty_weight': 1.0,
│   'batch_size': 4,
│   'open_in_gui': False,
│   'peak_threshold': 0.2,
│   'max_instances': None,
│   'tracking.tracker': None,
│   'tracking.target_instance_count': None,
│   'tracking.pre_cull_to_target': None,
│   'tracking.pre_cull_iou_threshold': None,
│   'tracking.post_connect_single_breaks': None,
│   'tracking.clean_instance_count': None,
│   'tracking.clean_iou_threshold': None,
│   'tracking.similarity': None,
│   'tracking.match': None,
│   'tracking.robust': None,
│   'tracking.track_window': None,
│   'tracking.min_new_track_points': None,
│   'tracking.min_match_points': None,
│   'tracking.img_scale': None,
│   'tracking.of_window_size': None,
│   'tracking.of_max_levels': None,
│   'tracking.save_shifted_instances': None,
│   'tracking.kf_node_indices': None,
│   'tracking.kf_init_frame_count': None
}

INFO:sleap.nn.inference:Auto-selected GPU 0 with 15098 MiB of free memory.
Versions:
SLEAP: 1.3.1
TensorFlow: 2.8.4
Numpy: 1.22.4
Python: 3.10.6
OS: Linux-5.15.109+-x86_64-with-glibc2.35

System:
GPUs: 1/1 available
  Device: /physical_device:GPU:0
         Available: True
        Initalized: False
     Memory growth: True

Predicting... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% ETA: -:--:-- ?
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/tensorflow/python/data/ops/iterator_ops.py", line 836, in __next__
    return self._next_internal()
  File "/usr/local/lib/python3.10/dist-packages/tensorflow/python/data/ops/iterator_ops.py", line 819, in _next_internal
    ret = gen_dataset_ops.iterator_get_next(
  File "/usr/local/lib/python3.10/dist-packages/tensorflow/python/ops/gen_dataset_ops.py", line 2923, in iterator_get_next
    _ops.raise_from_not_ok_status(e, name)
  File "/usr/local/lib/python3.10/dist-packages/tensorflow/python/framework/ops.py", line 7186, in raise_from_not_ok_status
    raise core._status_to_exception(e) from None  # pylint: disable=protected-access
tensorflow.python.framework.errors_impl.OutOfRangeError: End of sequence [Op:IteratorGetNext]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/sleap/nn/inference.py", line 435, in _predict_generator
    for ex in self.pipeline.make_dataset():
  File "/usr/local/lib/python3.10/dist-packages/sleap/nn/data/pipelines.py", line 287, in make_dataset
    ds = transformer.transform_dataset(ds)
  File "/usr/local/lib/python3.10/dist-packages/sleap/nn/data/normalization.py", line 342, in transform_dataset
    test_ex = next(iter(ds_input))
  File "/usr/local/lib/python3.10/dist-packages/tensorflow/python/data/ops/iterator_ops.py", line 838, in __next__
    raise StopIteration
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/bin/sleap-track", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.10/dist-packages/sleap/nn/inference.py", line 5412, in main
    labels_pr = predictor.predict(provider)
  File "/usr/local/lib/python3.10/dist-packages/sleap/nn/inference.py", line 526, in predict
    self._make_labeled_frames_from_generator(generator, data)
  File "/usr/local/lib/python3.10/dist-packages/sleap/nn/inference.py", line 1583, in _make_labeled_frames_from_generator
    for ex in generator:
RuntimeError: generator raised StopIteration

Thanks!

**Comment by roomrys** [comment]
Hi @jramborger78,

This looks like tensorflow is having trouble creating a dataset form your input "'colab.pkg.slp" which is the generic name used for the pkg.slp to predict on. I just want to verify that "colab.pkg.slp" is the name of the pkg.slp that you want to predict on (is this in the folder you exported in the training package?). Otherwise, do you mind changing that to the name of the pkg.slp that is in your training package folder?

Thanks,
Liezl

**Reply by jramborger78** [reply]
Yes, i set the name of the project to "colab" to make the unzip be "colab.pkg.slp".

**Reply by roomrys** [reply]
Hi @jramborger78,

Are you able to send the training package zip over via [this form](https://forms.gle/gwfjJ1KqkWVfwrQs8) for troubleshooting?

Thanks,
Liezl

**Reply by jramborger78** [reply]
Yes I can, thank you!

**Reply by roomrys** [reply]
Hi @jramborger78,

I have been troubleshooting your training package and also came across your same error, but when I create a new training package from my own data, the notebook seems to run without issue. Can you walk me through how you created the training package?

Thanks,
Liezl
